feat: trim dependencies + 318 PR

MrDwarf7 opened the pull request:
# Summary 


## Refactoring section 

- Removes the requirement for both `cfg-if` and `rand`, preferring  the use of (already available) cfg decorator for relevant function(s).
- Single branch logic in `erase_handler.rs` that was originally double-nested behind macro checks is now a single flat function call.
- `install_handler.rs` now handles the multiple repeat calls with the same `Err` returns via an inline async closure.
- The large stack of `Command::new(..)` calls for `git` actions are now handled in the `remote_checks(..)` function, which has also had a `#[rustfmt::skip]` macro applied to it to lessen the effect of large method chaining that's mostly non-relevant to understand the function call use-case.

## Cargo section

- Aforementioned removal of deps.
- Improved readability for `Cargo.toml` by using the standard syntax that almost all Rust projects use for declaring dependencies.
- Removed entirely redundant `optional = false` clauses from `Cargo.toml`
- Benchmarked difference on binary size and compilation time using `[profile.<profile>.package."*"]` and came to the conclusion that it's a marginal improvement to provide correct build instructions for project's dependencies - These each align with `debug`, `release`, and your custom one of `optimized` (for size).
- Moved `[[bin]]` block to top so other Rust developers can easily locate the required bin file, `binstall` calls remain at the bottom as they're unlikely to need to be changed or sought after.

## Benefits

- Lower binary size - we're now at 5.7M on the `optimized` profile.
- Due to the changes in #333  + the package opts, output feels smoother (I assume due to less buffering happening during the sys-call/request).

MordechaiHadad commented on Cargo.toml:
```diff
@@ -114,6 +73,24 @@ strip = true
 lto = true
 codegen-units = 1
 
+[profile.dev.package."*"]
```
Nvm i mixed it with profile.dev

MordechaiHadad commented on src/handlers/install_handler.rs:
```diff
@@ -642,6 +617,44 @@ async fn handle_building_from_source(version: &ParsedVersion, config: &Config) -
+    Ok(())
+}
+
+async fn metadata_handler<F, Fut>(dir: &str, action: F) -> Result<()>
```
This could be potentially removed and changed to Path::exists, i opted to use exists checks via metadata cuz sometimes the exists function didnt work correctly.

MordechaiHadad replied:
Oh damn i didnt notice the response to my comment here, fair.

MrDwarf7 committed:
feat: add cargo-make script(s) & formatting tools

ci/cd: move RUST_ settings to env

ci: actions-rs/cargo@v1 depr. use CLI directly

ci/cd: wrap name fields in quotes

ci: update path to prevent collision

refactor: erase/install use comptime

chore(deps): remove cfg-if and random from deps

Commit also contains generalized
cleanup of the Cargo.toml file
plus dep opts.
Improves readability and build times.

chore(help/proce): cleanup nested #cfg macros

Removes nesting and moves certain dependencies 
inside platform-specific cfg! macro blocks.


note: `if cfg!(..)` doesn't remove code at comp,
where as `#[cfg(..)] { ... }` does.

This is why the usage of it over the former

chore(Cargo.toml): remove comment, sort deps

ref(processes): remove loop & streamline

test(processes): add tests for is_neovim_running

chore(Cargo.lock): remove zerocopy in lockfile

chore: run formatters over the codebase

fix: test_is_version test failing due to change
fix: test_is_hash function due to change

chore: deepsource thinks the variable is empty

feat: remove requirement for cfg-if

MrDwarf7 committed:
chore: update deps (flate, tar)

MrDwarf7 committed:
chore: update dependencies

MrDwarf7 committed:
chore: update Windows deps (winreg, zip)

MordechaiHadad reviewed (commented):
Hmmm  I see you merged the separate PR with make and such with this, I am curious to know what is your reasoning behind adding taplo, cargo make, and all these files in build folder.

MrDwarf7 commented:
@MordechaiHadad - As you've noticed I like to keep PR's up to date in order to make merging as straightforward as possible. 
This was no exception with both 318 and this PR (338), however I found due to the nature the PR that manages formatting changes such as 338 (both codebase-wide and within the `Cargo.toml` file) it was becoming increasingly difficult to manage against the master branch changes happening.

The use of `taplo` is a `.toml` formatter, this maintains consistency across user contributions / PR's.

**Regarding `cargo make` and the files under the `./build/` folder** -

The way that `cargo make` works is effectively a `Makefile` without the headache of `Make` the language and the quirks it brings (such as tab spacing, and the language itself).

`Cargo make` is also 'extensible' which is what you're seeing in the `./build/` folder.
I keep things modular via the `extends` field in the `Makefile.toml` in the root directory.
It effectively acts as an 'entry' file and provides a handful of aliases. 

The `.toml` files under `./build/` are the `recipes` (as Make would call them). 
You can see that here using the `extend = [ { ... }]` toml array: 
https://github.com/MordechaiHadad/bob/pull/338/changes#diff-9375fd04332c86472d7be397ef09428cb86babd8826880a5835bd1d1c1bdbc08R1-R6

The other files under the `./build/` folder are: 
- **push_events.json**: This is used with [nektos/act](https://github.com/nektos/act) which is a GitHub Actions runner for locally testing actions, it's used as a 'mock' GitHub Actions runner without having to wait for the machines on GitHub's servers and leads to faster iterations when creating Actions files (and less commit noise / squashing required).
- **`.build.sh`**: This was originally used (on Unix systems) to prevent the nesting of release files and was called by either the `Makefile.toml` or via GitHub Actions (I've personally since in-lined the script anyway in my own GitHub Actions files)

Further (deeper) explanations on certain aspects of `cargo make` can be found in #318 itself or I'm happy to explain.

MordechaiHadad commented:
Hmm i see, thanks for the elaboration. I will try to rereview this PR its rather large.

MordechaiHadad commented on .github/workflows/cd.yml:
```diff
@@ -1,169 +1,139 @@
+          chmod +x linuxdeploy linuxdeploy-plugin-appimage
+      - name: "Prepare AppDir"
+        if: matrix.os.NAME == 'linux' && matrix.tls.NAME == 'Rustls'
+        run: "mkdir -p AppDir/usr/bin AppDir/usr/share/icons/hicolor/256x256/apps AppDir/usr/share/applications\ncp target/optimized/bob AppDir/usr/bin/\ncp resources/bob-icon.png AppDir/usr/share/icons/hicolor/256x256/apps/bob.png\ncat <<EOF > AppDir/bob.desktop\n[Desktop Entry]\nName=Bob Neovim Manager\nExec=bob\nIcon=bob\nType=Application\nCategories=Utility;Development;\nComment=A cross-platf ...
```
Why are you using this one liner with a bunch of \n?

MrDwarf7 replied:
I'm assuming this was the formatter doing this (likely misaligned I assume), I can fix this via copy-paste from original + fix whatever the formatter was seeing, good pickup.

MordechaiHadad replied:
I think we should try and use cargo appimage or whatever it's called to produce an appimage, my implementation is barely maintainable. I'll try to push it sometime soon before I merge this so we don't have to deal with this anymore

MordechaiHadad commented on rust-toolchain.toml:
```diff
@@ -0,0 +1,11 @@
+[toolchain]
+channel = "nightly"
```
Why do we use nightly here? I am not sure of an instance when i needed more than the stable toolchain for bob

MrDwarf7 replied:
I've used nightly practically forever, there's mostly optimizations (for both output of code & for comp. time's), that and I almost always use `cranelift` and a custom linker and `cranelift` requires nightly iirc.

MordechaiHadad replied:
Any reason to use it for the main environment though?

MrDwarf7 replied:
When you say:

> for the main environment

I'm not quite following what you mean by this? 

Setting the channel via the `rust-toolchain.toml` file will auto-download the toolchain if it's not already installed when building (Meaning if we set stable for instance, it'll build against stable for everything - ie: same for local development and for releases etc.)

Building against nightly ensures that we're always ahead of; and in support for, stable Rust versions due to how their release schedule works.

MordechaiHadad replied:
Yes exactly which is why you think we should use stable for everyone else in this case, nightly isn't considered cuz it can break after all, and I never needed to upgrade to nightly in order to work on this, i work fully on stable. So I am curious what are the benefits of moving to a less stable release.

MrDwarf7 replied:
One is that I use the `nightly` formatter as it's generally more optimized and changes for formatting are; for the most part, merged into stable if they make it into `nightly` regarding the formatter. 

One other example I can give on functionality is the Rust ecosystem transition for `rustls` vs. `rust-native-tls`/`rustls`, which I caught on a couple projects before it ended up having an effect on the main branch.

As I mentioned earlier - `nightly` in Rust isn't quite as much of a `nightly` as many other languages which I've used (I'm assuming that's largely down to the base language having such strict rules already and being a solid base to build up off of).
Unless we're using 'weird' or particularly niche features (Ones that require a crate-wide `#![feature(...)]` macro) or multi-layer crate features (Those specified in the `Crate.toml` file for your own crate) then we are _**highly**_ unlikely to run into those edge cases. Your depreciation of multi-ssl backends (via #382 ) is actually a good example of finding ways to remove comp-time feature flags that are no longer required that could have potential to cause issues anyway.

MordechaiHadad replied:
🤔hmmm okay will try to use nigthly on windows, we can roll back the toolchain anyhow if there are some issues.

MordechaiHadad commented on src/handlers/list_handler.rs:
```diff
@@ -177,12 +178,13 @@ mod list_handler_is_version_tests {
-                false => assert!(!is_version(case)),
-            });
+        cases_expected.iter().for_each(|(case, expected)| {
+            if *expected {
```
Why if clause over the previous match? is it automatic formatting?

MrDwarf7 replied:
This was from debugging, I found the per-line output easier to locate iirc as I'd had other things like `prinln!(...);`'s and `dbg!(...);) around it.
Either works as far as test cases go, though I do find the original I wrote more elegant, yes.

MordechaiHadad replied:
Yes I do prefer the original match case, 🤔 should be rolled back to it

MordechaiHadad commented on src/helpers/version/mod.rs:
```diff
@@ -322,12 +326,13 @@ mod version_is_hash_tests {
-                true => assert!(is_hash(case)),
-                false => assert!(!is_hash(case)),
-            });
+        version_expected.iter().for_each(|(case, expected)| {
```
Same question as above regarding the if clause over match statement

MordechaiHadad commented:
A few more comments
1. Regarding security_audit can you upgrade the version for flate
2. if there are any questions you have regarding deepsource sometimes odd analysis let me know (i should probably make it explicit in contributing file)

MrDwarf7 commented:
Yeah, I can fix up the commented items and address the issue with `flate` later today, should be good to go then too.

MrDwarf7 committed:
test(list_handler,verison::mod): prefer match

MrDwarf7 commented:
@MordechaiHadad - These changes should be complete now, I've rebased over the `dev` branch and fixed any conflicts.

- `cd.yml` fixed (directly from `dev` branch as of d4569ab48091b4592053701300f95151644b9959)
- `list_handler.rs` and its test case has been updated/reverted to use the original `match` statement.
- `version::mod.rs` also updated/reverted to use original `match` statement.

I've not yet updated the `rust-toolchain.toml` file we have an on-going conversation about it.

When you're reviewing this, please double-check against the intended outcome regarding deps. as merging these is a big pita due to `Cargo.lock` diffs and the branch we're basing off now also has other changes (`tls` and such).

MrDwarf7 committed:
chore(unarchive): collapse `if` statement

MrDwarf7 committed:
test(ci/cd): use nightly Rust in CI/CD workflows

Note: this is mostly a test push to
test against the formatting job under
.github/workflows/ci.yml

MrDwarf7 committed:
fix(version::mod): ref fn to pass deepsource prof.

MrDwarf7 committed:
ref(get_version_sync_file_location): return Option<PathBuf> instead of
Result<Option<PathBuf>>

MrDwarf7 committed:
ci/cd: use env. level OUTPUT_ARTIFACT_DIR + specifcy shell for `run:`
calls

MrDwarf7 committed:
ref(get_version_sync_file_location): use generic <P> where P:
AsRef<Path>

MrDwarf7 committed:
chore(install_handler): use const for env check

MrDwarf7 committed:
chore(processes): document public function `is_neovim_running`

MrDwarf7 committed:
chore(install_handler): streamline progress bar update logic

MrDwarf7 committed:
ci/cd: 'copy' cmd should be `cp` via bash shell

MrDwarf7 commented:
Ahhh, now I remember the other reason I changed the `match *expected` from a match to a `if-else` block.
Deepsource has a fit over it ( [RS-C1011](https://app.deepsource.com/gh/MordechaiHadad/bob/run/38d9c5dd-490b-4b63-b66c-04a7665f9c63/rust/RS-C1011?listindex=0) ) and says it's an 'anti-pattern'.

Either you can disable that rule or I can try and figure out a way to re-write it with something like `.map(...)` or `.fold(...)` but at this point we're fiddling with a test case that runs either way just to appease the "Great DeepSource" which isn't really worth it.

The other one is the cyclomatic complexity lint it's picking up for a function no one's touched in a long time at this stage anyway, and it's already marked with a clippy annotation re: `#[allow(clippy::too_many_lines)]`.
Does DeepSource itself have a comment annotation or something that can be used to tell it to ignore certain things? Similar to `#[rustfmt::skip]` or `#[allow(clippy::XX)]`?

MrDwarf7 commented:
Yep, there we go, now it's passing all of them, so we're good to go.

MordechaiHadad commented:
> Ahhh, now I remember the other reason I changed the `match *expected` from a match to a `if-else` block. Deepsource has a fit over it ( [RS-C1011](https://app.deepsource.com/gh/MordechaiHadad/bob/run/38d9c5dd-490b-4b63-b66c-04a7665f9c63/rust/RS-C1011?listindex=0) ) and says it's an 'anti-pattern'.
> 
> Either you can disable that rule or I can try and figure out a way to re-write it with something like `.map(...)` or `.fold(...)` but at this point we're fiddling with a test case that runs either way just to appease the "Great DeepSource" which isn't really worth it.
> 
> The other one is the cyclomatic complexity lint it's picking up for a function no one's touched in a long time at this stage anyway, and it's already marked with a clippy annotation re: `#[allow(clippy::too_many_lines)]`. Does DeepSource itself have a comment annotation or something that can be used to tell it to ignore certain things? Similar to `#[rustfmt::skip]` or `#[allow(clippy::XX)]`?

Ah okay.... deepsource is a bit annoying sometimes lmao i will disable it 
and yes skipcq in comments 
https://deepsource.com/blog/releases-silence-issues-in-code

MordechaiHadad commented on Cargo.toml:
```diff
@@ -1,123 +1,95 @@
+futures-util          = { version = "0.3.14" }
+indicatif             = "0.18.0"
+regex                 = "1.5"
+reqwest               = { version = "0.12", features = ["stream", "rustls-tls-native-roots"] } ## cannot upgrade as feature = "rustls-tls-native-roots" was removed in 0.13.*
```
Oh right yes I forgot about that, I think 0.13 now uses rustls directly as default with `rustls-platform-verifier` so you can just set to 0.13, and pass rustls directly just to be more explicit.

MrDwarf7 replied:
This has been updated/completed (There's also a note in the commit that changes this too for anyone that might do dev goes back).

MordechaiHadad commented on src/handlers/install_handler.rs:
```diff
@@ -510,111 +483,80 @@ fn file_type_ext(version: &ParsedVersion, get_sha256sum: bool) -> std::borrow::C
 
     // create neovim-git if it does not exist
     let dirname = "neovim-git";
-    if let Err(error) = fs::metadata(dirname).await {
```
A little note here on why I use metadata to check if it doesnt exist is because there was some bug a few years earlier when Path::exists() sometimes decided to work and sometimes didnt so using this metadata hack was the most reliable at the time

MordechaiHadad replied:
I just realized i said this in oct 17

MordechaiHadad commented:
Not sure why the CI shows pending checks when the CI clearly finished, but it seems currently it fails to upload the built binary to the CI pipeline artifacts.

MordechaiHadad commented on build/.build.sh:
Hmmm why do we need this if dist-tasks.toml exists?

MrDwarf7 replied:
Because it's in my template that I use regularly and, at the time I believe you had an external script too. You can find the one sitting in my public template repo [here](https://github.com/MrDwarf7/rust_template/blob/main/build/.old_build.sh)

MordechaiHadad replied:
> Because it's in my template that I use regularly and, at the time I believe you had an external script too. You can find the one sitting in my public template repo [here](https://github.com/MrDwarf7/rust_template/blob/main/build/.old_build.sh?rgh-link-date=2025-12-31T11%3A04%3A18Z)

I see... but what exactly is the purpose of it existing? I don't see it referenced in the CI or cargo make tasks

MrDwarf7 replied:
I can remove it, it was previously the externalized script (of what is now inline in my own GitHub Actions files), happy to remove it.

MordechaiHadad replied:
Ah alright, if you think it can help someone in local dev env keep it otherwise yeah better to remove it

MrDwarf7 replied:
There's certainly potential for it to be useful in local development settings (ie: hand-in-hand with `act`), but for now I've removed it to get this finalized.

MrDwarf7 commented:
The reason, more or less that actions hang is more often than note [this](https://youtu.be/E3_95BZYIVs?si=e2nqLO9NhJ7P78Rr), feel free to watch it. There may be other reason but at this point I'll happily chalk it up to GitHub Actions being janky

MrDwarf7 committed:
chore: remove old build file

MrDwarf7 committed:
ci/cd(Makefile): add `bo`, `ba`, and `ro` recipes.

- `bo` - Buld Optimized
- `ba` - Build All (While new, it also now includes bulding optimized)
- `ro` - Run Optimized

MrDwarf7 committed:
chore: update the `reqwest` crate to 0.13.1

The dep. update addresses a breaking change now that
the default TLS handling is done via `rustls` instead of 
`native-tls`.

Dev Note:
Associated internal functions provided by reqwest also had many
changes in this release bump, along with removal of far-deprecated
options and public API functions.

MrDwarf7 committed:
docs(CONTRIBUTING): add `cargo-make` to documentation.

The addition of `cargo-make` to the project also requries
relevant 'on-boarding' documentation, which has been added 
to the `CONTRIBUTING.md` file.
It outlines how to install `cargo-make` and `taplo-cli` pre-reqs,
provides 2 examples of usage, and points the developer to 
the `Makefile.toml` in the root of the project for other 
aliases/recipes.

MrDwarf7 commented:
I've cleaned up the prior request & also added relevant recipes for the `optimized` builds.
Additionally, I've also updated the `CONTRIBUTING.md` file to outline the usage of `cargo-make` along with relevant install instructions for both Linux and Windows systems.

MordechaiHadad commented:
> I've cleaned up the prior request & also added relevant recipes for the `optimized` builds. Additionally, I've also updated the `CONTRIBUTING.md` file to outline the usage of `cargo-make` along with relevant install instructions for both Linux and Windows systems.

Thanks, btw I am moving this PR to be merged into dev I think we should target this branch for most future PRs

MordechaiHadad commented:
I see that the windows' artifact is still not being uploaded correctly and is just populated with the cargo make toml scripts, can you check whats up with that? I also see that the artifacts sizes have jumped by quite a lot, like my PR resulted in macos arm to be 1.97MB and this one to 2.71MB, could you run cargo-bloat and see what might be the issue?

MrDwarf7 committed:
ci/cd: use maintained version of checkout action

Additionally: wrapped all name fields in quotes
because yaml is a great data fmt language

MrDwarf7 committed:
ci/cd: use YAML anchors for OUTPUT_ARTIFACT_DIR

MrDwarf7 commented:
Alright that's the `Windows` build not being copied into the correct folder resolved, as for the size, I'll have to look at exactly why this is. I have a feeling it may be to do with the `reqwest` update and bundling of their new TLS.

MordechaiHadad commented:
> Alright that's the `Windows` build not being copied into the correct folder resolved, as for the size, I'll have to look at exactly why this is. I have a feeling it may be to do with the `reqwest` update and bundling of their new TLS.

Awesome let me know, feel free to dump cargo bloat output here if needed.

MrDwarf7 commented:
From what I can see after running it and quick checking yesterday, it's the aws encryption crate (I can see 2 in the bloat output) which is taking up majority of the size (Perhaps different versions, so we build with 2 internally at link). I imagine when whatever other project we have moves up a dep. major this will resolve.

I can look further but I know the deps. we hold currently can't really be trimmed further as it is anyway. I can test reverting the dep. version and see what size the builds come out to.

As an aside: I am curious (though understanding, at least from my own pov) why we're so careful about binary size, and consequently - what you consider an 'upper limit' to size?

MordechaiHadad commented:
Regarding your question about being careful with binary size, it's because I just prefer lower binary sizes. I prefer to take up as little space as possible on the user's computer. Regarding upper limit there's no real limit, I just try to balance experience with size, I won't for example take a massive crate which gives us a 1% benefit for the same reason I won't rule out a crate that is maybe heavier but will improve the DX/UX.

MordechaiHadad commented on CONTRIBUTING.md:
```diff
@@ -34,20 +36,80 @@ Thanks for wanting to contribute! Bob is community-driven and we appreciate all
 
-# Build
+# Build using cargo-make
+makers a # or cargo-make a
```
Is makers basically a default alias that comes with cargo make command?

MrDwarf7 replied:
Correct (I have it aliases to `ma` personally).

MordechaiHadad commented on build/push_event.json:
Not sure exactly what is this file used by

MrDwarf7 replied:
This can be used to locally simulate GH Action runners via [act](https://github.com/nektos/act) - which I also provided a link to here: https://github.com/MordechaiHadad/bob/pull/338#issuecomment-3673136604 when you asked about the `./build/` folder.

MordechaiHadad replied:
Ahhh my bad missed it, good idea.

MordechaiHadad commented on .github/workflows/cd.yml:
```diff
@@ -117,40 +93,44 @@ jobs:
+      - name: "Setup Bob build directory"
+        env:
+          OUTPUT_ARTIFACT_DIR: *output_artifact_dir
+        shell: bash
```
Bash on windows?

MrDwarf7 replied:
The runner doesn't actually care haha, always bound it amusing too.

MordechaiHadad replied:
Ah I see, please delete it, it is a bit confusing at first glance.
